### PR TITLE
Issue #3950: NoWhitespaceAfter: add support for method reference operator

### PR DESCRIPTION
**Pull request by Kietzmann**
PR related to issue #3950

### CODE REVIEW

File: config/checkstyle_checks.xml [hunk 1]
```diff
@@ -470,6 +470,7 @@
       <property name="tokens" value="UNARY_PLUS"/>
       <property name="tokens" value="ARRAY_DECLARATOR"/>
       <property name="tokens" value="INDEX_OP"/>
+      <property name="tokens" value="METHOD_REF"/>
```
Comment (rnveach) [inline config/checkstyle_checks.xml]: @romani Please confirm you are ok with this token here.

Comment (romani) [inline config/checkstyle_checks.xml]: I agree

**Comment by romani** [comment]
@Kietzmann , please run reports generation to make sure that no obvious problem on real code.

**Comment by Kietzmann** [comment]
@romani 
I'll do it ASAP

**Comment by Kietzmann** [comment]
@romani @rnveach 
could you please tell me what config.xml to use to run?

**Comment by romani** [comment]
On all projects, you need to uncomment them.
Config.xml is where you define what Check should be used by checkstyle please read http://checkstyle.sourceforge.net/cmdline.html .Et us know if still not clear.

**Comment by Kietzmann** [comment]
In my case, I had modified checkstyle_checks.xml and I need to pass it to the diff.groovy
so I do 
groovy diff.groovy --localGitRepo /home/dkytsmen/IdeaProjects/checkstyle --baseBranch master --patchBranch 3950 --config /home/dkytsmen/IdeaProjects/checkstyle/config/checkstyle_checks.xml --listOfProjects projects-to-test-on.properties
I have uncommented all projects in projects-to-test-on.properties file
When I run it, I am recieving 
Failed during checkstyle configuration: unable to parse configuration stream: Property ${checkstyle.suppressions.file} has not been set -> [Help 1]

For the launch.groovy I suppose that I should make:
mvn clean install -DskipTests -Dcobertura.skip=true -Dfindbugs.skip=true -Dpmd.skip=true
and the run :
groovy launch.groovy --listOfProjects projects-to-test-on.properties --config /home/dkytsmen/IdeaProjects/checkstyle/config/checkstyle_checks.xml

Am I right?

**Comment by rnveach** [comment]
> In my case, I had modified checkstyle_checks.xml and I need to pass it to the diff.groovy

Go into `checkstyle-tester` folder, and modify `my_check.xml` with your check and **ONLY** your check in place of `ThrowsCount` example in there by default.
You can specify your check however you want it with any/all options. Proof enough that your change is valid and doesn't break or have any weird/bad behaviors.

> Failed during checkstyle configuration: unable to parse configuration stream: Property ${checkstyle.suppressions.file} has not been set

This is why you can't use checkstyle's configuration file and must make your own.

> For the launch.groovy 

Don't use launch at all, only `diff.groovy`.

**Comment by Kietzmann** [comment]
@rnveach 
Should I run it on all projects in the properties file, or just in single or few?

**Comment by rnveach** [comment]
@Kietzmann Read what romani wrote.

> On all projects, you need to uncomment them.

**Comment by Kietzmann** [comment]
@rnveach 
The last question)
When I am running
groovy diff.groovy -r /home/dkytsmen/IdeaProjects/checkstyle -p 3950 -c my_check.xml -l projects-to-test-on.properties
It seems that git switches to master branch and I'm receiving 
 Token "METHOD_REF" was not found in Acceptable tokens list in check com.puppycrawl.tools.checkstyle.checks.whitespace.NoWhitespaceAfterCheck 
Seems that it use file version from master branch
What should I modify to run it correctly?
Thanks

**Comment by rnveach** [comment]
> Token "METHOD_REF" was not found in Acceptable tokens list in check

@Kietzmann I am sorry, I forgot this PR was adding a new token that is not in default list.
We are in talks to change `diff.groovy` to be able to run regression on new tokens which it doesn't support right now.

You need to use `launch.groovy` in the meantime.
Sorry for the confusion.

**Comment by Kietzmann** [comment]
@romani @rnveach 
please review
kietzmann.github.io

**Comment by romani** [comment]
404

File not found

**Comment by Kietzmann** [comment]
https://kietzmann.github.io/apache-ant/site/index.html
And so on for other project reports

**Comment by romani** [comment]
Please make index page to let us know other links

**Comment by rnveach** [comment]
@Kietzmann All your reports are missing the Java files.

https://kietzmann.github.io/openjdk9/site/index.html
https://kietzmann.github.io/openjdk9/site/xref/openjdk9/src/jdk.internal.le/share/classes/jdk/internal/jline/TerminalFactory.html#L156

**Comment by Kietzmann** [comment]
@rnveach 
And how to run it to contain Java files?
I ran
groovy launch.groovy -l projects-to-test-on.properties -c my_check.xml
with my_chek.xml containing

     <module name="NoWhitespaceAfter">
     <property name="tokens" value="METHOD_REF"/>
      </module>

in  in place of ThrowsCount

**Comment by Vladlis** [comment]
Report should contain them by default. Please, check that you share exactly what is generated.

**Comment by Kietzmann** [comment]
@Vladlis @rnveach 
I ran groovy diff.groovy -r /home/dkytsmen/IdeaProjects/checkstyle -p 3950 -c my_check.xml -l projects-to-test-on.properties
with following files
https://gist.github.com/Kietzmann/52cc0c36bbe2dcdc279d5a65c32a3a31
https://gist.github.com/Kietzmann/6f7b55ff9051366e288b42010a5a1039

Then I get reports directory and pushed it on github

**Comment by Kietzmann** [comment]
@Vladlis @rnveach 
What should I do to fix it?

**Comment by rnveach** [comment]
> I ran groovy diff.groovy 

I assume you mean `launch.groovy`.

Looking at your repository at https://github.com/Kietzmann/Kietzmann.github.io/tree/master/openjdk9/site/xref , it looks like you have the xref directory, but no sub-folders in it.
@Kietzmann 
What OS are you using?
Can you go to `openjdk9`'s xref directory on your local and see if you see any subfolders in it?

**Comment by Kietzmann** [comment]
@rnveach 
Yes it was mistyped.
I'm running Ubuntu 14.04, Groovy Version: 2.4.8
in `reports/openjdk9/site/xref` I haven't any subfolder

**Comment by rnveach** [comment]
@MEZk @Kietzmann I believe I see the issue. We had a similar issue with our older regression.
https://github.com/Kietzmann/Kietzmann.github.io/blob/master/openjdk9/site/index.html#L71
All your results are on one line and it is causing the links to not be identified correctly and thus is removing all your xref files. We don't know why this happens on some machines and not others.
I will create an issue for this.

**Comment by Kietzmann** [comment]
@rnveach 
I'll try to run this script on my home laptop with Ubuntu
Maybe it will have different behavior
Or run it on Windows?

**Comment by romani** [comment]
@Kietzmann , it should work on both, just make it on both to speedup the process.

**Comment by Kietzmann** [comment]
@rnveach @romani 
The same result I'm receiving on Ubuntu and Windows on laptop: no reports for exact java files in every xref subfolder. Could you please help me to complete this PR with running regression report?
Thanks

**Comment by romani** [comment]
@Kietzmann , please share full log of launch by gist.github.com, please share with us your contribution fork where all your changes are present to let us run the same command on the same configuration that you have.

**Comment by rnveach** [comment]
@Kietzmann What is the result of your `System.getProperty( "line.separator" )` in java?

**Comment by Kietzmann** [comment]
@romani 
Is there any log file generated by launch.groovy or I should rerun it and copy logs from console?

**Comment by Kietzmann** [comment]
@rnveach 
It is `

`

**Comment by romani** [comment]
Redirect output to file and share with us

**Comment by Kietzmann** [comment]
@romani 
https://gist.github.com/Kietzmann/66bc7cca113d81ca725ec2f7f7758f78

**Comment by romani** [comment]
please share a link to your contribution repo (with all your changes)

what I see suspicious for now is space is path `/media/dmytro/New Volume/` .  Shell language is fragile for this. @Kietzmann , please try to use another path (without spaces).
You suppose to run groovy (but now I am not sure what is launches).

Please share all details:
code that you run
command that you run
output (already provided)

**Comment by Kietzmann** [comment]
@romani 
I tried another path without spaces on my desktop
the result was the same
command:
groovy launch.groovy --listOfProjects projects-to-test-on.properties --config my_check.xml

repo:
https://github.com/Kietzmann/checkstyle/tree/3950

**Comment by romani** [comment]
i need fork of your contribution repo, from where you run groovy to verify that you all did right

**Comment by Kietzmann** [comment]
@romani 
https://github.com/Kietzmann/contribution/tree/3950

**Comment by Kietzmann** [comment]
@romani 
Is there anything I could do to speed up the process?

**Comment by MEZk** [comment]
@Kietzmann 
We introduced [the fix](https://github.com/checkstyle/contribution/commit/a640cd6c570d5d892d9f893248a29cba8fdd0fb6) for your problem. Please, update your local fork of Checkstyle contribution repo and rerun the script.

**Comment by Kietzmann** [comment]
@MEZk 
Thanks
Updating and starting

**Comment by rnveach** [comment]
travis and wercker are common errors we are getting right now. they can be ignored.

reports look good to me.

**Comment by romani** [comment]
Travis is restarted.
As #4330 merged, please rebase to resolve wercker problem

**Comment by Kietzmann** [comment]
@romani 
I've rebased with main already.

**Comment by Kietzmann** [comment]
@romani 
Make review and merge

**Pull request merged**